src/agent/agent.c

```diff
@@ -1842,12 +1842,6 @@ static int agent_method_switch_controller(sd_bus_message *m, void *userdata, UNU
                 return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Failed to set CONTROLLER ADDRESS");
         }
 
-        r = sd_bus_emit_properties_changed(
-                        agent->api_bus, BC_AGENT_OBJECT_PATH, AGENT_INTERFACE, "ControllerAddress", NULL);
-        if (r < 0) {
-                bc_log_errorf("Failed to emit controller address property changed: %s", strerror(-r));
-        }
-
         bc_log_infof("CONTROLLER ADDRESS changed to %s", dbus_address);
 
         agent_disconnected(NULL, userdata, NULL);
```

Comment:
Instead of the changes in `agent_reconnect`, we could also move emitting the signal after the `agent_disconnect` here:
```suggestion
        agent_disconnected(NULL, userdata, NULL);
        
        r = sd_bus_emit_properties_changed(
                        agent->api_bus, BC_AGENT_OBJECT_PATH, AGENT_INTERFACE, "ControllerAddress", NULL);
        if (r < 0) {
                bc_log_errorf("Failed to emit controller address property changed: %s", strerror(-r));
        }
```
This keeps the code a bit simpler and should result in the new value for `ControllerAddress` being emitted. It also leads to the agent having tried to connect to the controller at least once before emitting the signal.
What do you think about this? @dofmind

Reply:
I was worried about to emit the signal being delayed but your suggestion seems more reasonable. I will update.

Reply:
Good point about the delay. Thinking more about the implications a bit more... the sequence of signals is off (first we'll receive the node status changes and then the signal for the address change).
What do you think about directly calling `agent_set_orch_address(agent, dbus_address)` right after `agent_set_controller_address(agent, dbus_address)` and keep emitting the `ControllerAddress` before disconnecting ?

Reply:
Calling `agent_set_orch_address(agent, dbus_address)` seems to be a workaround. When `agent->controller_address` is not NULL, how about using `agent->controller_address` instead of `agent->orch_addr` for the property?

```
diff --git a/src/agent/agent.c b/src/agent/agent.c
index febbba4..e93d740 100644
--- a/src/agent/agent.c
+++ b/src/agent/agent.c
@@ -1921,6 +1921,23 @@ static int agent_property_get_log_target(
         return sd_bus_message_append(reply, "s", log_target_to_str(bc_log_get_log_fn()));
 }
 
+/*************************************************************************
+ **** org.eclipse.bluechi.Agent.ControllerAddress *****
+ *************************************************************************/
+static int agent_property_get_controller_address(
+                UNUSED sd_bus *bus,
+                UNUSED const char *path,
+                UNUSED const char *interface,
+                UNUSED const char *property,
+                sd_bus_message *reply,
+                void *userdata,
+                UNUSED sd_bus_error *ret_error) {
+        Agent *agent = userdata;
+        const char *address = agent->controller_address ? agent->controller_address : agent->orch_addr;
+
+        return sd_bus_message_append(reply, "s", address);
+}
+
 
 static const sd_bus_vtable agent_vtable[] = {
         SD_BUS_VTABLE_START(0),
@@ -1939,8 +1956,8 @@ static const sd_bus_vtable agent_vtable[] = {
                         SD_BUS_VTABLE_PROPERTY_EXPLICIT),
         SD_BUS_PROPERTY("ControllerAddress",
                         "s",
-                        NULL,
-                        offsetof(Agent, orch_addr),
+                        agent_property_get_controller_address,
+                        0,
                         SD_BUS_VTABLE_PROPERTY_EMITS_CHANGE),
         SD_BUS_VTABLE_END
 };
```

Reply:
The `ControllerAddress` property should really only point to the used address - which is `orch_addr` (naming is a bit off, it should probably refactored later). 

Since the signal should be emitted when `orch_addr` changes, the right place to emit the signal should probably be the setter function `agent_set_orch_address` - and remove emitting the signal from the switch controller function. This is a bit similar to your initial change, but would lead to the signal being emitted every time `orch_addr` changes (e.g. when the host name get resolved differently). Something like this:
```C
bool agent_set_orch_address(Agent *agent, const char *address) {
        bool should_emit = (agent->orch_addr == NULL && address != NULL) ||
                        (agent->orch_addr != NULL && address != NULL && !streq(agent->orch_addr, address));

        bool result = copy_str(&agent->orch_addr, address);
        if (should_emit && result) {
                int r = sd_bus_emit_properties_changed(
                                agent->api_bus, BC_AGENT_OBJECT_PATH, AGENT_INTERFACE, "ControllerAddress", NULL);
                if (r < 0) {
                        bc_log_errorf("Failed to emit controller address property changed: %s", strerror(-r));
                }
        }
        return result;
}
```
What do you think? @dofmind The setter gets more complex, but probably worth it.

Reply:
Since `agent_set_orch_address` is just a setter function, I am not sure if it is correct to emit a signal from the setter function. Also, there may be some corner cases here. If setting `agent->orch_addr` to NULL and then calling `agent_set_orch_address` with the same address , a signal will be emitted even though `agent->orch_addr` has not actually changed.

Setting `agent->orch_addr` is caused by `agent_reconnect` and `agent_set_orch_address`. Currently `agent_set_orch_address` is only called by `ensure_orch_address` (the `main` of agent also calls `agent_set_orch_address`, but only once in the initial phase. And I think it should be changed to `agent_set_controller_address`). `ensure_orch_address` is called by `agent_start` and `agent_reconnect`. Since `ensure_orch_address` is called before `agent->api_bus` is set, there is no need to consider `agent_start`. So `agent_reconnect` might be the best place to emit signal for the property. Hmm how about consider my first changes again?

If my first changes are still not enough to merge, I'll update with your suggestion to emit a signal from the setter.
 `agent_method_switch_controller` handles the above corner case because it already checks for identical address settings.

Reply:
>  ensure_orch_address is called by agent_start and agent_reconnect. Since ensure_orch_address is called before agent->api_bus is set, there is no need to consider agent_start. So agent_reconnect might be the best place to emit signal for the property. 

Fair point, ok, lets go with your initial changes and add it to the `agent_reconnect`. 
Instead of calling `free(orch_addr)`, it makes sense to use `_cleanup_free_ char *orch_addr = NULL;` (which will take care of the cleanup). 
Could you restore your initial changes with that small adjustment? I think then we are ready to merge.

Reply:
Sure, i will update soon.